snapd: initial implementation for systemd software watchdog for snapd

mvo5 opened the pull request:
The CE team was asking for a software watchdog implementation for snapd in the teamcall we had yesterday. This is a first draft. If it looks reasonable I'm happy to add a proper spread test and more unit testing.

The bug in question: https://bugs.launchpad.net/snapd/+bug/1674506

zyga commented on data/systemd/snapd.service:
```diff
@@ -6,6 +6,7 @@ Requires=snapd.socket
 ExecStart=/usr/lib/snapd/snapd
 EnvironmentFile=/etc/environment
 Restart=always
+WatchdogSec=5m
```
The watchdog should be poked at twice the frequency specifies here from what I recall. How do we ensure that?

zyga commented on systemd/sdnotify.go:
```diff
@@ -0,0 +1,82 @@
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse WATCHDOG_USEC: %s", err)
+	}
+	dur := time.Duration(usec/2) * time.Microsecond
```
Aha :-)

zyga commented on systemd/sdnotify.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	raddr := &net.UnixAddr{
+		Name: e,
```
Perhaps heavyweight but how about a watchdog manager that opens th socket once at startup and handles this all internally?

mvo5 replied:
I was following the code in libsystemd/sd-daemon/sd-daemon.c - it uses this structure, i.e. open socket, send data and _cleanup_close_ the fd.

zyga commented:
This is failing on `[ 1235.846963] audit: type=1326 audit(1492083223.916:961): auid=1000 uid=0 gid=0 ses=3 pid=11968 comm="snap-exec" exe="/usr/lib/snapd/snap-exec" sig=31 arch=c000003e syscall=49 compat=0 ip=0x561dc6ddc1f4 code=0x0`. (bind)

chipaca commented on systemd/sdnotify.go:
```diff
@@ -0,0 +1,82 @@
+	if wu == "" {
+		return nil, fmt.Errorf("cannot get WATCHDOG_USEC environment")
+	}
+	usec, err := strconv.Atoi(wu)
```
Not sure it's wanted/helpful for this case, but note you have `osutil.GetenvInt64`.

chipaca commented on systemd/sdnotify.go:
```diff
@@ -0,0 +1,82 @@
+			select {
+			case <-wt.C:
+				sdNotify("WATCHDOG=1")
+			}
```
shouldn't `RunWatchdog` take a tomb, and check it here?
E.g, `func RunWatchdog(d *daemon.Daemon)`  and then `case <-d.Dying()`?

pedronis commented:
not a blocker but as  I mentioned a slightly more meaningful implementation would involve a new pair of (settable) timer or ticker/callback in the overlord loop

Overlord.SetHeartbeat(interval, cb)

or something like that

then the OverlordLoop would have (sketching):

    case <-heartbeatC:
         cb()
         continue ...

zyga commented on cmd/snapd/main.go:
```diff
@@ -64,6 +90,12 @@ func run() error {
+	if ticker, err := runWatchdog(d); err != nil {
+		logger.Noticef("cannot run software watchdog: %s", err)
+	} else {
+		defer ticker.Stop()
```
Hmm, am I missing something? If we fail you just log the message and if we don't fail you ... stop the ticker?

mvo5 replied:
Thanks! You are right about the notify, this is silly because if snapd fails to check-in with systemd, it will be killed anyway. The ticker is stopped in a defer (just like the daemon is stopped). But I chagned the code now to make it more obvious.

zyga commented on systemd/sdnotify.go:
```diff
@@ -0,0 +1,56 @@
+// inspired by libsystemd/sd-daemon/sd-daemon.c from the systemd source
+func SdNotify(state string) error {
+	if state == "" {
+		return fmt.Errorf("cannot use empty state")
```
The term `state` is misleading as it may imply the snapd state. Not sure how to call this to make it better though

mvo5 replied:
Thanks! I changed this now.

chipaca commented:
what's the state of this PR?

mvo5 commented:
I updated the PR and addressed the review feedback. It does not address the concerns from @pedronis yet, if we want this in this PR I think we should close and reopen only once it is reworked.

pedronis commented:
@mvo5 since I wrote my comment I noticed that especially because of auto-refresh code we might do in Ensure things that can take a fairly long time relatively speaking, we would need to bound them somehow precisely or move them out again to use the Overlord loop for the watchdog functionaility

mvo5 commented:
Given that we actually had no instance of when the system became unresponsive I will close this PR. The current form will not actually help much, all it proves is that a single go-routine in snapd is still alive which is not super helpful.
